behaviours/meme_rater.py

```diff
@@ -62,68 +62,83 @@ def number_emoji(number: int):
 
 
-async def msg_update(event: hikari.GuildMessageUpdateEvent) -> None:
-    if event.channel_id != MEME_CHANNEL_ID or not event.message.embeds:
-        return
+async def _process_attachment(
+    attachment: hikari.Attachment, username: str | None
+) -> agents.MemeAnswer | None:
+    """Process a single attachment and return rating and explanation."""
+    att_ext = (attachment.extension or "").lower()
+    if att_ext not in IMG_FILE_EXTENSIONS:
+        return None
+    return await get_meme_rating(attachment.url, username)
+
+
+async def _process_embed(
+    embed: hikari.Embed, username: str | None
+) -> agents.MemeAnswer | None:
+    """Process a single embed and return rating and explanation."""
+    image_url = None
+    if embed.thumbnail:
+        image_url = embed.thumbnail.proxy_url
+    elif embed.video and embed.video.proxy_url:
+        image_url = f"{embed.video.proxy_url}?format=webp&width={embed.video.width}&height={embed.video.height}"
+    if not image_url:
+        return None
+    return await get_meme_rating(image_url, username)
 
-    # We only want to rate the meme if it was not edited after the initial post
-    # Discord does not add an edited timestamp when embedding an image
-    if event.message.edited_timestamp:
-        return
 
-    ratings = list[int]()
-    explanations = list[str]()
-
-    for embed in event.message.embeds:
-        image_url = None
-        if embed.thumbnail:
-            image_url = embed.thumbnail.proxy_url
-        elif embed.video and embed.video.proxy_url:
-            image_url = f"{embed.video.proxy_url}?format=webp&width={embed.video.width}&height={embed.video.height}"
-        if not image_url:
-            continue
-
-        try:
-            username = None
-            if event.author:
-                username = event.author.username
-            res = await get_meme_rating(image_url, username)
-            if not res:
-                continue
-            ratings.append(res.rate)
-            explanations.append(res.explanation)
-        except Exception:
-            continue
-    await rate_meme(event.message, ratings, explanations)
+async def process_message_content(
+    message: hikari.PartialMessage,
+) -> tuple[list[int], list[str]]:
+    """Process all meme content in a message and return ratings and explanations."""
+    # Ensure we have the full message
+    if not isinstance(message, hikari.Message):
+        message = await message.app.rest.fetch_message(message.channel_id, message.id)
+
+    ratings: list[int] = []
+    explanations: list[str] = []
+
+    # Process attachments
+    for attachment in message.attachments:
+        attachment_rating = await _process_attachment(
+            attachment,
+            message.author.username,
+        )
+        if attachment_rating:
+            ratings.append(attachment_rating.rate)
+            explanations.append(attachment_rating.explanation)
+
+    # Process embeds
+    for embed in message.embeds:
```

Comment:
I previously commented on this to refactor the repeated loops. I can't see the comment now, though. Please action.

Reply:
Abstracted. But I'm out of ideas on how to do it declaratively or in one line.
The functions mutate 2 different list variables.

Reply:
```python
_ = [(ratings.append(er.rate), explanations.append(er.explanation)) for er in [await _process_embed(embed, message.author.username) for embed in message.embeds] if er]
```

Reply:
Not what I was thinking at all, and now I'd say this is worse than it was. The only difference between the two loops is `message.attachments` and `_process_attachments`. So, write a function that takes these two as parameters, and runs the loops.

A draft:
```
def foo(process_x, xs):
    answers = []
    for x in xs:
        answer = await process_x(x)
        if answer:
            answers.append(x)
    return xs 
```

I would also hope in the same refactoring that you'd refactor the parallel lists into lists of pairs/some other type

Reply:
@adamnwerbik, please don't close. This is not resolved

Reply:
2 loops are still used, but all Media URLs are first extracted from embeds & attachments then rated in one go.
And rating results are returned as a single list.

I hope this refactor matches your expectations.

Reply:
That looks much nicer. Two loops are unavoidable, but having them both reuse the same logic to extract a rating is great.